src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+    }
+    public static bool MyIsNullOrEmpty([NotNullWhenFalse] string? s, [EnsuresNotNull] string? s2) => throw null;
+}
+" + NotNullWhenFalseAttributeDefinition + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+
+            VerifyAnnotations(c, "C.MyIsNullOrEmpty", NotNullWhenFalse, EnsuresNotNull);
+        }
+
+        [Fact]
+        public void NotNullWhenFalse_AndEnsuresNotNull()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+public class C
+{
+    void Main(string? s)
+    {
+        if (MyIsNullOrEmpty(s))
+        {
+            s.ToString(); // ok
+        }
+        else
+        {
+            s.ToString(); // ok
+        }
+
+        s.ToString(); // ok
+    }
+    public static bool MyIsNullOrEmpty([NotNullWhenFalse, EnsuresNotNull] string? s) => throw null;
+}
+" + NotNullWhenFalseAttributeDefinition + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+
+            VerifyAnnotations(c, "C.MyIsNullOrEmpty", NotNullWhenFalse | EnsuresNotNull);
+        }
+
+        [Fact]
+        public void EnsuresNotNull_Simple()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+public class C
+{
+    void Main(string? s)
+    {
+        ThrowIfNull(42, s);
+        s.ToString(); // ok
+    }
+    public static void ThrowIfNull(int x, [EnsuresNotNull] string? s) => throw null;
+}
+" + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+
+            VerifyAnnotations(c, "C.ThrowIfNull", None, EnsuresNotNull);
+        }
+
+        [Fact]
+        public void EnsuresNotNull_OnDifferentTypes()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+public class C
+{
+    public static void Bad<T>([EnsuresNotNull] int i) => throw null;
+    public static void ThrowIfNull<T>([EnsuresNotNull] T t) => throw null;
+}
+" + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (5,32): error CS8628: The 'EnsuresNotNullAttribute' attribute is not applicable on a value type.
+                //     public static void Bad<T>([EnsuresNotNull] int i) => throw null;
+                Diagnostic(ErrorCode.ERR_AttributeNotApplicableOnValueType, "EnsuresNotNull").WithArguments("EnsuresNotNullAttribute").WithLocation(5, 32)
+                );
+
+            VerifyAnnotationsInSource(c, "C.Bad", None);
+            VerifyAnnotationsInSource(c, "C.ThrowIfNull", EnsuresNotNull);
+        }
+
+        [Fact]
+        public void EnsuresNotNull_GenericMethod()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+public class C
+{
+    void Main(string? s)
+    {
+        ThrowIfNull(s);
```

Comment:
>ThrowIfNull(s); [](start = 8, length = 15)

Please add a test for `ThrowIfNull(t); t.ToString();` where t is an unconstrained type parameter.